From 8aae873f35907055795db00fa9962bfe047661d0 Mon Sep 17 00:00:00 2001
From: Vitaut Bajaryn <vitaut.bayaryn@gmail.com>
Date: Sun, 11 Feb 2018 01:33:00 +0100
Subject: [PATCH 2/2] ci#1 Mount child datasets of the root too

Note: this will also mount datasets that have canmount=off.
---
 defaults/linuxrc | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/defaults/linuxrc b/defaults/linuxrc
index 4f22217..5f3c381 100644
--- a/defaults/linuxrc
+++ b/defaults/linuxrc
@@ -954,6 +954,31 @@ else
 
 fi # if [ "${CDROOT}" = '1' ]
 
+
+if [ "${ROOTFSTYPE}" = 'zfs' ]; then
+       # mount child datasets of the root
+       # TODO take some better logic from initramfs-tools
+       zfs list -H -o name -r "${REAL_ROOT}" | while read i; do
+               suffix="${i#${REAL_ROOT}}"
+
+               if [ "$(zfs get -H -o value mountpoint ${i})" = 'legacy' ]
+               then
+                       MOUNT_STATE=rw
+               else
+                       MOUNT_STATE=rw,zfsutil
+               fi
+
+               if [ "${REAL_ROOTFLAGS}" = '' ]; then
+                       good_msg "Using mount -t ${ROOTFSTYPE} -o ${MOUNT_STATE} ${i} ${NEW_ROOT}${suffix}"
+                       mount -t ${ROOTFSTYPE} -o ${MOUNT_STATE} ${i} ${NEW_ROOT}${suffix}
+               else
+                       good_msg "Using mount -t ${ROOTFSTYPE} -o ${MOUNT_STATE},${REAL_ROOTFLAGS} ${i} ${NEW_ROOT}${suffix}"
+                       mount -t ${ROOTFSTYPE} -o ${MOUNT_STATE},${REAL_ROOTFLAGS} ${i} ${NEW_ROOT}${suffix}
+               fi
+       done
+fi
+
+
 # Mount the additional things as required by udev & systemd
 if [ -f ${NEW_ROOT}/etc/initramfs.mounts ]; then
 	fslist=$(get_mounts_list)
-- 
2.11.1

